udp-reassembly-offload.md

```diff
@@ -21,7 +21,12 @@ The Winsock API (currently only software fallback) already exists, and details o
 
 This section describes necessary updates in the Windows network stack to support URO.
 
-> **TODO**
+- TCPIP will disable the hardware offload, and the software fallback, if any WFP/LWF filters that are URO-incompatible are installed on the system.
+- If hardware URO is not available, but a socket opts-in to URO with the Winsock API, software URO will be used for that socket. Software coalescing will follow the same rules as NDIS, except for the following:
+    - Software coalescing may decide to use multiple NBLs in a chain of up to 255, instead of a single NBL.
+    - Software coalescing may also coalesce UDP payloads of different lengths using the above method.
```

Comment:
"Software coalescing may also coalesce UDP payloads of different lengths using the above method"

Did I misunderstand anything or are we going to change our current URO implementation? Our URO implementation bails if it sees a NB's data length is different than the one in the SCU.

Reply:
Yes, I am thinking of changing the existing URO implementation to support more scenarios than hardware. I'll have to see how many places require this assumption to gauge the work involved.

Reply:
But why? It doesn't seem like it's very useful to me. And to implement it, you will have to keep track of the boundaries of all packets coalesced somewhere (not in the OOB info field). If a sender does not send full mss packets, I don't really see a reason to coalesce them. It's bound to not have good throughput.

Reply:
The size can be kept track of in the NBL. Since it's just a chain of NBLs, each one keeps track of its length.  Fair point that this might not occur in frequent enough situations to be useful.

Reply:
I don't really understand this software URO you are describing. What's really get coalesced?

Reply:
Again, we should apply the simplest is best strategy and just go with what we know and keep things the same.

Reply:
Ok. I'll remove this for now and we can consider this in the future if it seems like it might offer benefit with real-world data.